libs/utils/energy_model.py

```diff
@@ -0,0 +1,106 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2016, ARM Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+from collections import namedtuple, OrderedDict
+
+import pandas as pd
+import numpy as np
+
+from devlib.utils.misc import memoized
+
+ActiveState = namedtuple("ActiveState", ["capacity", "power"])
+ActiveState.__new__.__defaults__ = (None, None)
+
+class EnergyModelNode(namedtuple("EnergyModelNode",
+                                 ["cpus", "active_states", "idle_states",
+                                  "power_domain", "freq_domain"])):
+    @property
+    def max_capacity(self):
+        return max(s.capacity for s in self.active_states.values())
+
+    def idle_state_idx(self, state):
+        return self.idle_states.keys().index(state)
+
+EnergyModelNode.__new__.__defaults__ = (None, None, None, None, None)
+
+class PowerDomain(object):
+    def __init__(self, idle_states, parent, cpus):
+        self.cpus = set()
+        self.idle_states = idle_states
+
+        self.parent = parent
+        self.add_cpus(set(cpus))
+
+    def add_cpus(self, cpus):
+        self.cpus = self.cpus.union(cpus)
+        if self.parent:
+            self.parent.add_cpus(self.cpus)
+
+    def __repr__(self):
+        return "PowerDomain(cpus={})".format(list(self.cpus))
+
+class EnergyModel(object):
+    """
+    Represents hierarchical CPU topology with power and capacity data
+
+    Describes a CPU topology similarly to trappy's Topology class, additionally
+    describing relative CPU compute capacity, frequency domains and energy costs
+    in various configurations.
+
+    The topology is stored in "levels", currently hard-coded to be "cpu" and
+    "cluster". Each level is a list of EnergyModelNode objects. An EnergyModel
```

Comment:
Yeah, that's a good point. I think probably they should have no name at all and just be numerically indexed. We can assume that level 0 is the logical CPU. I'll play around with that and see if it works.